### DISCUSSION TITLE
Welcome to eventuous discussions!

### DISCUSSION THREAD

**Discussion by alexeyzimarev** [Eventuous/eventuous#2] [General]
## 👋 Welcome!

We’re using Discussions as a place to connect with other members of our community. We hope that you:
* Ask questions you’re wondering about.
* Share ideas.

Remember that you can find answers to some of your questions on the documentation at [eventuous.dev](https://eventuous.dev).

To get started, comment below with an introduction of yourself and tell us about what you do with this community.

If you want to discuss something, open a new discussion. Do not use comments for existing discussions for new questions.

**Comment by shashikantpawar21** [comment]
Myself - Shashikant Pawar 
Trying to use eventuous in a project

**Comment by shashikantpawar21** [comment]
Hi Alexey, we are planning to use eventuous for our production... Could you let us know about the license the project has ?

**Reply by shashikantpawar21** [reply]
Hey Alexey ...anything on above request ?

**Reply by JulianMay** [reply]
It says 'MIT' in the read.me

**Reply by shashikantpawar21** [reply]
Thank you JulianMay for a quick help... Are you also maintainer of the library ?

**Reply by alexeyzimarev** [reply]
The next version will change to Apache 2.0, not sure if this will be a problem.

**Reply by shashikantpawar21** [reply]
That is fine ..Thank you for the info

**Reply by JulianMay** [reply]
@shashikantpawar21 nope, just thought I'd chime in since I knew the answer

**Comment by shashikantpawar21** [comment]
@alexeyzimarev @JulianMay 
In the shared example ... BookingState will store all the data in memory use to return on request of api.. Just wanted to know in production if we are creating many Booking..let say 1000 booking per day. Will this be a issue ?

**Reply by JulianMay** [reply]
It's not stored, just returned - however you manage a response-cache or anything like that (in-memory snapshots etc.) is up to you, and your own maintenance/monitoring responsibility.
But I'd say that used as-is even 1000 a minute should not be a problem on modern infrastructure.

**Reply by shashikantpawar21** [reply]
Thank you Julian..

**Reply by bartelink** [reply]
@shashikantpawar21 I know you are only being friendly, but can I suggest avoiding atting people in github discussions unless you are providing specific warranted follow-ups to questions being asked please - we need to appreciate and conserve the time of maintainers of OSS by letting them answer asychronously as and when they have time.
If you're paying or sponsoring, or have been asked sepcifically to @ then that may be a different matter of course.

**Reply by alexeyzimarev** [reply]
Thanks @bartelink 

I am not sure where the idea about storing anything in memory comes from. Eventuous doesn't store anything in memory. We also have a production system processing 1000 commands per _second_ without any issues.

**Reply by JulianMay** [reply]
To maintainers: can you advise me if I'm actually being unhelpful by answering queries of others here in 'discuss'? My intention is to help, not to overstep 😌
My OSS etiquette is not evolved, don't know the unwritten rules. Thanks in advance!

**Reply by bartelink** [reply]
@JulianMay IMO you're absolutely not - you've been helpful throughout, and taking the load off maintainers is a big thing people like you and I can do to let maintainers focus on higher value things. (Or at least that's what I try to do in the discussions and/or issues for serilog and xunit with this intent)

The main point of my intervention on this thread was to point out the (most likely unintended) effect that atting someone has - they get notifications right away and there is an implicit onus on them to answer things. In other words, the fact that someone has answered a question in the past may well lead you to think they will know the answer to the next question too, but let them discover the question in their own time, if and when they next have it!

**Reply by alexeyzimarev** [reply]
Julian, I really appreciate your responses!

**Reply by JulianMay** [reply]
Thanks, I'll stick around then!

**Reply by shashikantpawar21** [reply]
Apologies for atting people..

**Comment by shashikantpawar21** [comment]
My Aggregate class has to handled many command and its related handler. As the class is getting bigger so i tried using partial keyword to split the logic between multiple files. It does not work good for methods to be partial

**Reply by JulianMay** [reply]
This _could_ be a smell that the aggregate boundary is not great (too big), typically this happens when aggregates are modelled around "a set of data" instead of "behaviour around a set of shared invariants". 
That being said, you _could_ always delegate to event-sourced value-objects and entities _inside_ the aggregate, and have the (aggregate) root handle the cross-entity/value validation and delegation.
What I mean in practical terms is to have entities/values (separate classes/records) which have their own method that returns events for the root to 'AddChange'  (or takes an action-argument to do so).

**Reply by shashikantpawar21** [reply]
Thank you Julian... That's helped.

**Comment by shashikantpawar21** [comment]
Difficult to understand when to create a aggregates... I created 4 aggregate based on sub-domain in my system. 
Out of these 4,  3 aggregate needs to know mostly all messages of remaining one aggregate so it seems to me as main aggregate. 
Should i passed the message between aggregates through Shovel or RabbitMQ.. which is better ?
Or i am missing something ?

**Reply by JulianMay** [reply]
Some constraints are hard, some are soft - it's a modelling problem more than a technical one.
Maybe relax some constraints, maybe coordinate using a saga/process manager (maybe that process is the _real_ aggregate, and the rest is just 'data'?), maybe go for choreography and have individual policies/eventhandlers changing some aggregates based on events of others.

In essence, you cannot have atomic transactions across streams with a database like EventStore (and for good reasons).

Take a look at https://ieeexplore.ieee.org/document/1407829
and https://codeopinion.com/avoiding-distributed-transactions-with-the-reservation-pattern/ for inspiration.

**Reply by shashikantpawar21** [reply]
Thank you for the information.. I will have a look on shared articles

**Reply by shashikantpawar21** [reply]
I also feel that it is modelling problem ..Can you share me your thought on below scenario.. my problem is similar to it.

Order service - changes order state from create, processed to shipped 
All the other service are independent but need to know about the state changes of order service as they constraint on it.. 
Like payment service, customer service ..so on.. 
This means every time order state changes ..order service sends a event and all other service subscribe to it and keep a note in their memory about it...

Am i doing something wrong here or any better way to handle it ?

**Reply by JulianMay** [reply]
That sounds fine and reasonable to me. The constraints upon the 'order status' for the contexts/services downstream from 'OrderService' would need to accept _some_ lag and the possibility/risk that the status might have changed a millisecond earlier without it knowing about it yet (when handling their own commands). This is a situation you will never be in if everything is in the same monolithic sql-database, but also the very thing that makes the approach scale way way beyond the single sql-database. Furthermore, the sql-database doesn't help consistency in the real world, which works more like a distributed event processing system than a centralised ACID-transactional DB - things happen!

This 'risk' is a very normal thing in everyday business, for example: "_A customer canceled the order just as shipping is collecting it from the warehouse_". Whomever put's the products in the truck will need to keep their eyes away from a screen for longer than it takes for the status-change to propagate from order-service to shipping-service via rabbitMQ. 
The cancelled order can be shipped anyway, regardless of how instant the information is there - what remains is a business decision: what to do about orders that are cancelled after it's shipped? A lot of these kinds of risks are mostly theoretical and never actually happens, others happens once in a while but can easily be managed manually. Some happens frequently or have great business-impact, and in those cases it makes sense to automate what the business-policy for that case is. An example of this is Amazon sending a mail to the customer, appologizing that the product is out of stock (even though they thought it wasn't when the costumer received their order-cinfirmation), and refunding the customers money. It's typically quite easier to identify (and react) to "the thing has happened" than to _categorically_ prevent "the thing happening" with regards to constraints with dependencies from "outside" some boundary.

**Reply by shashikantpawar21** [reply]
Thank you for your detailed guidance..

**Comment by shashikantpawar21** [comment]
Do anyone have some sample code on how to use Eventuous.RabbitMq for sending messages between services and updating state

**Reply by JulianMay** [reply]
Look at the tests. RabbitMqProducer is used like so: https://github.com/Eventuous/eventuous/blob/dev/src/RabbitMq/test/Eventuous.Tests.RabbitMq/SubscriptionSpec.cs

Be sure only to publish (integration)-events based on internal (domain)-events _that have been committed to the eventstore_. The simples/cleanest way to do this, IMO, is to have the an internal event-handler subscribe to domain events from EventStoreDB and publish to rabbitMQ.

https://eventuous.dev/docs/subscriptions/

**Reply by alexeyzimarev** [reply]
The easiest way to do it today is using a gateway. You'd then only need a transformer. https://github.com/Eventuous/connector-esdb-elastic/blob/main/src/Eventuous.Connector.EsdbElastic/Startup.cs

**Reply by alexeyzimarev** [reply]
The usual scenario for handling commands would be:

- Produce commands using RMQ producer in one place
- In another place, have a subscription to the given exchange with one event handler, which inherits from the `EventHandler` abstract class
- The handler would take a command service as a dependency
- Add `On<DoThat>` handler to the event handler, which will just pass through the command to the command service.

**Reply by shashikantpawar21** [reply]
Thank you Julian May and Alexey for the help with information... 
I will try to do something with above details. 

Alexey ... I request you to cover below 2 scenario in sample code at https://github.com/Eventuous/dotnet-sample/
1. integration event through above gateway approach.
2. Distributed transaction through saga orchestration in event sourced system

**Reply by shashikantpawar21** [reply]
I did tried the Shovel thing mentioned for service communication but it did add PaymentsIntegration stream and i am trying to avoid it somehow

**Reply by shashikantpawar21** [reply]
Harder to understand with Rabbit MQ test and document are not available .. Is there any code for passing events to other services just like dotnet sample code. It would be really helpful

**Comment by shashikantpawar21** [comment]
Alexey ...
Fix the event store tracing --> https://github.com/Eventuous/eventuous/commit/3c9ec1a6a4213b3da47c8220e539462c5f03a20e
Does this commit means, we need to upgrade the eventuous version to latest one and what's the latest version ?
Is it stable one ?

**Reply by alexeyzimarev** [reply]
The package is alpha, new alpha packages get out several times a week. Alpha packages are not stable, and not available on nuget.org.

**Comment by shashikantpawar21** [comment]
**Scenario**
Two microservice handling order and product separately 
1. Order State --> Order-Stream-123   (OrderCreated, OrderCompleted)
2. Product State --> Product-Stream-123 ( ProductAdded , ProductRemoved) - product added inside order 123.

I want to create a Stream so i know what happens to product inside order
eg  OrderProductStream-123 --> OrderCreated, ProductAdded, ProductAdded, OrderCompleted

**Query** - How the above scenario to be handled with eventuous  as one state is mapped to one stream/aggregator

**Reply by alexeyzimarev** [reply]
I can only advise you to read my book about the basic DDD tactical patterns and aggregates design.

**Comment by shashikantpawar21** [comment]
How to delete a record from projection (mongo document)  with  UpdateDefinitionBuilder  methods
I did tried Pull, PullFilter and unset method but it does not work

**Comment by alexeyzimarev** [comment]
Pull and PullFilter are operations for updates. If you want to delete a document, you need to execute a delete operation. For that purpose, you can return an instance of `CollectionOperation<T>`

https://github.com/Eventuous/eventuous/blob/dev/src/Mongo/src/Eventuous.Projections.MongoDB/README.md (scroll to the end)

**Comment by shashikantpawar21** [comment]
I am trying to have vertical slice architecture so I have a case where I wanted to extend Aggregate class and i was able to do so but i have access to State property and but not Apply method in the new extension method as Apply method is protected. 

Is it harmful to make it protected internal or any other way to solve the issue?

**Reply by JulianMay** [reply]
Keeping the Aggregate(Root) contained in a single class seems very intentional by design. Maybe reading up on the basics behind the concepts (in this case 'aggregate root as a transactional boundary') would help you understand the deliberate constraints eventous enforces?

**Reply by JulianMay** [reply]
In other words - the aggregate protects invariants _across_ commands. For vertical architecture, the command and command-handler can be sliced per usecase, the aggregate needs to be re-used across commands with shared invariants

**Reply by JulianMay** [reply]
.. And thus commands that share no invariants probably should not belong to the same aggregate. A very common "mistake" with aggregate modelling is grouping aggregates by how you would group tables in a relational database, leading to big aggregates with awkward lifecycles

**Reply by alexeyzimarev** [reply]
I'd like to know the reason behind the motivation for extending the `Aggregate` base class as I can't find one that is legit. I'd rather prefer to have it sealed.

**Comment by shashikantpawar21** [comment]
Not sure why Type Map throw exception for some other library use in the application. 
Can you help me ignore or solve this error ?

![image](https://user-images.githubusercontent.com/83539291/171010787-33323f38-a64a-461d-86cd-4ebbe54f50d1.png)

**Reply by alexeyzimarev** [reply]
I have no idea, sorry.

**Reply by shashikantpawar21** [reply]
no worries . I figured out that the dll was missing and added it but not sure why the TypeMap function looks for it

**Comment by shashikantpawar21** [comment]
update.SetOnInsert().Set(x => x.Images, imageBytes)

I am able to use this function if my mongo columns is byte[]... It looks for string only 

Could you suggest anything on this ?

**Reply by alexeyzimarev** [reply]
I am sorry but I am unable to provide any help with using MongoDB.

**Comment by shashikantpawar21** [comment]
I have one event AddPhotosToAlbums { Guid albumId , List<string> photoFilePath }

While updating mongo projection i want add specific record for each photo to retrieve faster., 
But seems the UpdateDefinition method support only single insert in the table with update.SetOnInsert or AddToSet methods 

Any help on this ?

**Reply by alexeyzimarev** [reply]
You need to use this overload: https://github.com/Eventuous/eventuous/blob/f33d83cdbb943616699572bf3407e7d6a018f18c/src/Mongo/src/Eventuous.Projections.MongoDB/MongoProjection.cs#L29

and return a `CollectionOperation` instance from the handler function.

```csharp
        On<V1.PaymentRecorded>(
            evt => new CollectionOperation<MyBookings.Booking>(
                (collection, ct) => collection.UpdateManyAsync(.......your update here.....)
            )
        );
```
